### ISSUE TITLE
[Feature Request] Add parse/format support to string

### ISSUE THREAD

**Issue by fzgem18** [move-language/move#727]
# 🚀 Feature Request
## Motivation

`string.move` doesn't support the conversion from/to `MoveValue`, which would be useful.

## Pitch

**Describe the solution you'd like**
new functions
```
public fun fmt_utf8<T>(v: &T): String;
public fun parse<T>(s: &String): T;
```
new native functions like
```
native public fun internal_fmt_utf8<T>(v: &T): vector<u8>;
native public fun internal_parse<T>(v: &vector): T;
```

**What I want to discuss**
gas parameters of the new natives

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
Yes

**Comment by vgao1996** [comment]
@fzgem18 thanks for putting up the feature request! I see that you are trying to add two things with this proposal:
1. String formatting: this is definitely something Move currently lacks and I think implementing `public fun fmt_utf8<T>(v: &T): String;` would be a great first step. Let's work together come up with a more detailed design. cc @wrwg @tnowacki @sblackshear 
2. A generic parse function: this function would allow you to create a struct belonging to an arbitrary module, which is a fatal violation of Move's safety invariants so I don't think we can get it added in the form you suggested. Would you mind elaborating on what problem you are trying to solve with this? Maybe we could find an alternative solution to that.

**Comment by sblackshear** [comment]
One thing we could consider here is exposing the functionality that powers `debug::print` (which I think is essentially what a user would want from `fmt_utf8`).

**Comment by banool** [comment]
@vgao1996 Some maybe useful context is this came up after a variety of similar request came up in #dev-discussions, where people would request "how do I go from x type to a string" or "from a string to x type". Not that Move has traits, but in those terms, what people are reaching for is a Display and FromStr trait, for all types where you'd expect default implementations to exist in other languages.

Naturally that's a lot for just one PR so maybe this PR could just be scoped down to the "x to string" case, and not the other way around.

**Comment by fzgem18** [comment]
Thanks to all for the comments. agree with @banool.
What I requested is only the function that parses string to the primitive types of move. and I thought this is an essential stuff in a language. I am not sure what kinda fatal violation may happen yet, as I am a learner of move. could you please explain more details?

**Comment by vgao1996** [comment]
> only the function that parses string to the primitive types of move. and I thought this is an essential stuff in a language.

If the conversion is limited to primitive types (but not user defined types) then it's a totally different story. I fully welcome such parsing functions and agree that they are essential to the language.

I'd suggest we divide this proposal into two parts:
1. Implementing `fun fmt_utf8`. This should be relatively easy as we could reuse the code we have for `debug::print` as @sblackshear mentioned.
2. Coming up with conversions from string to primitive types. 
  a. This part may need a bit more design & discussion (e.g. whether to have individual conversion functions for each type)

**Comment by fzgem18** [comment]
I did some commits on the issue, particular to `fmt_utf8` function. Could you kindly review it?

**Comment by fzgem18** [comment]
Exactly, the name of the function should be meaningful. `string::value_to_string` or `string::from_value` would be familiar to the functionality and purpose. As @sblackshear mentioned, I thought we should use the functionality that powers `debug::print`, which is available in testing environment only, not just the function. So it should be `Display` trait implementation for `ValueImpl`.

**Comment by wrwg** [comment]
Sam meant this [debug print](https://github.com/move-language/move/blob/06f190eaad933ccfd697d9fa77b916d22130b6b7/language/move-stdlib/nursery/sources/debug.move#L5).
